components/supervisor/pkg/supervisor/docker.go

```diff
@@ -0,0 +1,260 @@
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"golang.org/x/xerrors"
+
+	"github.com/gitpod-io/gitpod/common-go/analytics"
+	"github.com/gitpod-io/gitpod/common-go/log"
+	"github.com/gitpod-io/gitpod/common-go/process"
+	"github.com/gitpod-io/gitpod/supervisor/api"
+	"github.com/gitpod-io/gitpod/supervisor/pkg/activation"
+	"github.com/gitpod-io/gitpod/supervisor/pkg/dropwriter"
+	"github.com/gitpod-io/gitpod/supervisor/pkg/terminal"
+)
+
+// if exit error happens one after another within dockerStartErrorBurstDuration then we're in the burst
+// and should stop trying after maxBurstDockerStartAttempts
+const (
+	maxBurstDockerStartAttempts   = 15
+	dockerStartErrorBurstDuration = time.Minute * 1
+	maxIntervalBetweenDockerStart = 15 * time.Second
+
+	logsDir             = "/workspace/.gitpod/logs"
+	dockerUpLogFilePath = logsDir + "/docker-up.log"
+)
+
+func socketActivationForDocker(parentCtx context.Context, wg *sync.WaitGroup, term *terminal.Mux, cfg *Config, telemetry analytics.Writer, notifications *NotificationService) {
+	defer wg.Done()
+
+	ctx, cancel := context.WithCancel(parentCtx)
+	defer cancel()
+
+	logFile, err := openDockerUpLogFile()
+	if err != nil {
+		log.WithError(err).Error("docker-up: cannot open log file")
+	} else {
+		defer logFile.Close()
+	}
+
+	var notificationSent bool
+	var notificationDisabled bool
+	notifyDockerUpFail := func() error {
+		if notificationDisabled || notificationSent {
+			return nil
+		}
+		notificationSent = true
+		defer func() {
+			notificationSent = false
+		}()
+		openLogs := "Open Logs"
+		dontShowAgain := "Don't Show Again"
+		resp, err := notifications.Notify(ctx, &api.NotifyRequest{
+			Level:   api.NotifyRequest_ERROR,
+			Message: "Docker daemon is failing to start.",
+			Actions: []string{openLogs, dontShowAgain},
+		})
+		if err != nil {
+			return err
+		}
+		telemetry.Track(analytics.TrackMessage{
+			Identity: analytics.Identity{UserID: cfg.OwnerId},
+			Event:    "gitpod_activate_docker_fail_notification",
+			Properties: map[string]interface{}{
+				"instanceId":     cfg.WorkspaceInstanceID,
+				"workspaceId":    cfg.WorkspaceID,
+				"debugWorkspace": cfg.isDebugWorkspace(),
+				"action":         resp.Action,
+			},
+		})
+		if resp.Action == dontShowAgain {
+			notificationDisabled = true
+			return nil
+		}
+		if resp.Action != openLogs {
+			return nil
+		}
+		gpPath, err := exec.LookPath("gp")
+		if err != nil {
+			return err
+		}
+		gpCmd := exec.CommandContext(ctx, gpPath, "open", dockerUpLogFilePath)
+		gpCmd.Env = childProcEnvvars
+		gpCmd.Stdout = os.Stdout
+		gpCmd.Stderr = os.Stderr
+		return gpCmd.Run()
+	}
+
+	burstAttempts := 0
+	backoffStrategy := backoff.NewExponentialBackOff()
+	backoffStrategy.MaxInterval = maxIntervalBetweenDockerStart
+	var lastExitErrorTime time.Time
+
+	for ctx.Err() == nil {
+		err := listenToDockerSocket(ctx, term, cfg, telemetry, logFile)
```

Comment:
I thought we discussed in the. morning to recreate listener completely on any error to reject all stale incoming requests, if we move it back it will stale again

or I misunderstand something?